go/k8s/controllers/user_controller.go

```diff
@@ -100,66 +142,144 @@ func (r *UserReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
 	defer cancel()
 
 	var (
-		err           error
-		backendURL    string
-		userID        int64
-		userState     UserState
-		userStatus    backendv1.UserStatus
-		retrievedUser *backendv1.User
+		err              error
+		backendURL       string
+		userID           int64
+		trustedDeveloper bool
+		userState        UserState
+		userStatus       backendv1.UserStatus
+		retrievedUser    *backendv1.User
 	)
 
 	defer func() {
 		if err != nil {
 			userStatus.State = err.Error()
 		}
 		if retrievedUser != nil {
+			if r.almondConfig.EnableDeveloperBackend && trustedDeveloper {
+				retrievedUser.Spec.Mode = "developer"
+			} else {
+				retrievedUser.Spec.Mode = "shared"
+			}
+			r.Update(ctx, retrievedUser)
 			retrievedUser.Status = userStatus
 			r.Status().Update(ctx, retrievedUser)
 			r.Log.Info("update status:", "user", retrievedUser.Spec.ID, "status", retrievedUser.Status)
 		}
 		r.Log.Info("--- end ---")
 	}()
 
-	backendURL, userID, err = r.getBackendURL(ctx, req.NamespacedName.Name, req)
+	userID, trustedDeveloper, err = r.getUserFromName(req.Name)
 	if err != nil {
-		return ctrl.Result{RequeueAfter: 2 * time.Second}, err
+		return ctrl.Result{}, err
 	}
+
+	if r.almondConfig.EnableDeveloperBackend && trustedDeveloper {
```

Comment:
I'm not sure I fully understand the logic here, but it seems that the logic for developer user and the logic for shared users is separate, except for the call to runEngine. Maybe we should separate Reconcile into two functions entirely.